File: ros2_ws/src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

class NumberCounterNode : public rclcpp::Node
{
public:
    NumberCounterNode() : Node("add_two_ints_client")
    {
        // You can improve this using a thread pool
        // thread_ = std::thread(std::bind(&AddTwoIntsClientNode::callbackAddTwoIntsService, this, 1, 4));
        threads_.push_back(std::thread(std::bind(&NumberCounterNode::callbackAddTwoIntsService, this, 1, 4)));
        threads_.push_back(std::thread(std::bind(&NumberCounterNode::callbackAddTwoIntsService, this, 2, 6)));
    }

private:
    // Create a thread to aboud stopping the loop
    // std::thread thread_;
    std::vector<std::thread> threads_;

    void callbackAddTwoIntsService(int a, int b)
    {
        auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "waiting for the server to be up...");
        }

        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request -> a=a;
        request -> b=b;

        auto future = client -> async_send_request(request);

        // It will block until we have a response or exception
        try
        {
            auto response = future.get();
            RCLCPP_INFO(this -> get_logger(), "%d + %d = %ld", a,b, response -> sum);

        }
        catch(const std::exception& e)
        {
            RCLCPP_ERROR(this -> get_logger(), "Service call feiled");
        }
        

    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2_ws/src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from functools import partial

from example_interfaces.srv import AddTwoInts

class AddTwoIntsClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        self.call_add_two_ints_servers(6, 7)

    def call_add_two_ints_servers(self, a, b):
        client = self.create_client(AddTwoInts, "add_two_ints")
        while not client.wait_for_service(
            1.0
        ):  # The 1.0 is to wait for service (1 second)
            self.get_logger().warn("Waiting for Server Add Two Ints...")

        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        future = client.call_async(request)  # This is kind a promise in Python
        future.add_done_callback(partial(self.callback_call_add_two_ints, a=a, b=b))

    def callback_call_add_two_ints(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(f"{str(a)} + {str(b)} = {str(response.sum)}")
        except Exception as e:
            # El marcador %r se utiliza para representar el valor de una variable o expresión utilizando su representación con repr().
            # Si solo pusieras (e) sin la coma, Python lo interpretaría como una expresión entre paréntesis y no como una tupla.
            self.get_logger().error("Service call failed %r" % (e,))


def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClientNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
